tensorflow/lite/experimental/kernels/gru_cell.cc

```diff
@@ -0,0 +1,97 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/lite/experimental/kernels/gru_cell.h"
+
+#include <limits>
+#include <vector>
+
+#include "tensorflow/lite/kernels/internal/optimized/optimized_ops.h"
+
+namespace tflite {
+namespace ops {
+namespace experimental {
+namespace gru_cell {
+
+using reference_ops::Concatenation;
+using optimized_ops::FullyConnected;
+using optimized_ops::ArrayMap;
+using optimized_ops::MapAsArrayWithLastDimAsRows;
+
+void GruCell(const RuntimeShape& input_shape, const float* input,
+             const RuntimeShape& state_shape, float* state,
+             const RuntimeShape& gate_weight_shape, const float* gate_weight,
+             const RuntimeShape& gate_bias_shape, const float* gate_bias,
+             const RuntimeShape& candidate_weight_shape,
+             const float* candidate_weight,
+             const RuntimeShape& candidate_bias_shape,
+             const float* candidate_bias, const RuntimeShape& output_shape,
+             float* output, const RuntimeShape& activation_shape,
+             float* activation, const RuntimeShape& concat_shape,
+             float* concat) {
+  const int n_batch = input_shape.Dims(0);
+  const int n_input = input_shape.Dims(1);
+  const int n_output = state_shape.Dims(1);
+
+  // [x h] = concat(input, state)
+  std::vector<float const*> concat_arrays_data;
+  std::vector<RuntimeShape const*> concat_arrays_shapes;
+  concat_arrays_data.push_back(input);
+  concat_arrays_data.push_back(state);
+  concat_arrays_shapes.push_back(&input_shape);
+  concat_arrays_shapes.push_back(&state_shape);
+  tflite::ConcatenationParams concat_params;
+  concat_params.axis = 1;
+  concat_params.inputs_count = concat_arrays_data.size();
+  Concatenation(concat_params, &(concat_arrays_shapes[0]),
```

Comment:
if the concatenation can give you better performance or easier implementation, sure, go for it.

if not, I think [x h] * gate_weight + gate_bias can be computed separately (gate_weight & gate_bias split can be done outside the loop), this can save you the overhead for concat & extra memory.

Reply:
right, it's really up to you, I'm not sure about GruBlockCellV2, but since you're doing concat inside the loop, this maybe not be best for performance.

but again, it's up to you